django/db/backends/mysql/operations.py

```diff
@@ -177,13 +177,45 @@ def sql_flush(self, style, tables, sequences, allow_cascade=False):
         # NB: The generated SQL below is specific to MySQL
         # 'TRUNCATE x;', 'TRUNCATE y;', 'TRUNCATE z;'... style SQL statements
         # to clear all tables of all data
+        schema_name = self.connection.settings_dict['NAME']
+
         if tables:
+            with self.connection.cursor() as cursor:
+                cursor.execute("""
+                SELECT table_name, table_rows, auto_increment
+                FROM information_schema.tables
+                WHERE table_schema = %s;
+                """, (schema_name,))
+                rows = cursor.fetchall()
+
             sql = ['SET FOREIGN_KEY_CHECKS = 0;']
-            for table in tables:
-                sql.append('%s %s;' % (
-                    style.SQL_KEYWORD('TRUNCATE'),
-                    style.SQL_FIELD(self.quote_name(table)),
-                ))
+
+            for table_name, table_rows, auto_increment in rows:
+                if table_name not in tables:
+                    continue
+
+                if table_rows <= 1000:
```

Comment:
That seems like an arbitrary heuristic?

Reply:
Yes. The following is quoted from the ticket.

> TransactionTestCase uses sql_flush, via the flush management command, to reset the database. Most test suites don't use very large tables, so using DELETE FROM should normally be faster in this case. Optimizing it MySQL to use DELETE FROM for flushing, at least for tables up to a certain size (1000 rows or similar heuristic) could save lot of time when using TransactionTestCase.
> https://code.djangoproject.com/ticket/31275

It sounds reasonable for me.